core/commands/refs.go

```diff
@@ -274,43 +286,53 @@ func (rw *RefWriter) writeRefsRecursive(n ipld.Node) (int, error) {
 	return count, nil
 }
 
-func (rw *RefWriter) writeRefsSingle(n ipld.Node) (int, error) {
-	c := n.Cid()
-
-	if rw.skip(c) {
-		return 0, nil
+// visit returns two values:
+// - first indicates if we should keep traversing the DAG.
+// - second indicates if the given Cid should be printed to the user.
+//
+// visit will do branch pruning depending on rw.MaxDepth, previously visited
+// cids and whether rw.Unique is set. i.e. rw.Unique = false and
+// rw.MaxDepth = 1 disables any pruning. But setting rw.Unique to true will
+// prune already visited branches.
+func (rw *RefWriter) visit(c *cid.Cid, depth int) (bool, bool) {
+	overMaxDepth := rw.MaxDepth >= 0 && depth > rw.MaxDepth
+
+	// When not unique, only avoid exploring too deep
+	// We do not track a set of visited nodes in this case.
+	// We do not print anything too deep though.
+	if !rw.Unique {
+		return !overMaxDepth, overMaxDepth
 	}
 
-	count := 0
-	for _, l := range n.Links() {
-		lc := l.Cid
-		if rw.skip(lc) {
-			continue
-		}
-
-		if err := rw.WriteEdge(c, lc, l.Name); err != nil {
-			return count, err
-		}
-		count++
+	// Unique == true.
+	// We want Unique Cids. We keep track of seen Cids, and their depth
+	if rw.seen == nil {
+		rw.seen = make(map[string]int)
 	}
-	return count, nil
-}
 
-// skip returns whether to skip a cid
-func (rw *RefWriter) skip(c *cid.Cid) bool {
-	if !rw.Unique {
-		return false
-	}
+	key := string(c.Bytes())
+	oldDepth, ok := rw.seen[key]
 
-	if rw.seen == nil {
-		rw.seen = cid.NewSet()
+	// Never explore over max-depth. Never print nodes over
+	// max depth.
+	if overMaxDepth {
+		return false, false
```

Comment:
Ditto about exploring nodes *at* max depth.

Reply:
Also, shouldn't this check be higher up? No need to do anything if we're over the max depth.